lib/format.c

```diff
@@ -0,0 +1,470 @@
+		if (*p != '}')
+			spec->type = *p++;
+	}
+	if (*p != '}')
+		return -EINVAL;
+	*pos = p+1;
+	return 0;
+}
+
+struct xbps_fmt *
+xbps_fmt_parse(const char *format)
+{
+	struct strbuf buf = {0};
+	const char *pos = format;
+	struct xbps_fmt *fmt = NULL;
+	size_t n = 0;
+	int r = 1;
+
+	for (;;) {
+		struct xbps_fmt_spec spec;
+		struct xbps_fmt *tmp;
+		r = nexttok(&pos, &buf);
+		if (r < 0)
+			goto err;
+		tmp = realloc(fmt, sizeof(*fmt)*(n + 1));
+		if (!tmp)
+			goto err_errno;
+		fmt = tmp;
+		switch (r) {
+		case 0:
+			fmt[n].common = NULL;
+			goto out;
+		case TTEXT:
+			fmt[n].chunk = calloc(1, sizeof(struct chunk)+buf.len+1);
+			fmt[n].common->type = TTEXT;
+			if (!fmt[n].chunk)
+				goto err_errno;
+			memcpy(fmt[n].chunk->s, buf.mem, buf.len+1);
+			break;
+		case TVAR:
+			r = parse(&pos, &buf, &spec);
+			if (r < 0)
+				goto err;
+			fmt[n].var = calloc(1, sizeof(struct var)+buf.len+1);
+			if (!fmt[n].var)
+				goto err_errno;
+			fmt[n].common->type = TVAR;
+			fmt[n].var->spec = spec;
+			memcpy(fmt[n].var->s, buf.mem, buf.len+1);
+			break;
+		}
+		n++;
+	}
+out:
+	strbuf_release(&buf);
+	return fmt;
+err_errno:
+	r = -errno;
+err:
+	free(fmt);
+	strbuf_release(&buf);
+	errno = -r;
+	return NULL;
+}
+
+void
+xbps_fmt_free(struct xbps_fmt *fmt)
+{
+	if (!fmt)
+		return;
+	for (struct xbps_fmt *f = fmt; f->common; f++)
+		switch (f->common->type) {
+		case TTEXT: free(f->chunk); break;
+		case TVAR: free(f->var); break;
+		}
+	free(fmt);
+}
+
+int
+xbps_fmts(const char *format, xbps_fmt_cb *cb, void *data, FILE *fp)
+{
+	struct strbuf buf = {0};
+	const char *pos = format;
+	int r = 0;
+
+	for (;;) {
+		struct xbps_fmt_spec spec;
+		r = nexttok(&pos, &buf);
+		if (r <= 0)
+			goto out;
+		switch (r) {
+		case TTEXT:
+			fprintf(fp, "%s", buf.mem);
+			break;
+		case TVAR:
+			r = parse(&pos, &buf, &spec);
+			if (r < 0)
+				goto out;
+			r = cb(fp, &spec, buf.mem, data);
+			if (r != 0)
+				goto out;
+			break;
+		}
+	}
+out:
+	strbuf_release(&buf);
+	return r;
+}
+
+int
+xbps_fmt(const struct xbps_fmt *fmt, xbps_fmt_cb *cb, void *data, FILE *fp)
+{
+	int r;
+	for (const struct xbps_fmt *f = fmt; f->common; f++) {
+		switch (f->common->type) {
+		case TTEXT:
+			fprintf(fp, "%s", f->chunk->s);
+			break;
+		case TVAR:
+			r = cb(fp, &f->var->spec, f->var->s, data);
+			if (r != 0)
+				return r;
+			break;
+		}
+	}
+	return 0;
+}
+
+struct fmt_dict_cb {
+	xbps_dictionary_t dict;
+};
+
+int
+xbps_fmt_string(const struct xbps_fmt_spec *spec, const char *str, size_t len, FILE *fp)
+{
+	if (len == 0)
+		len = strlen(str);
+	if (spec->align == '>' && spec->width > (unsigned int)len) {
+		for (unsigned i = 0; i < spec->width - len; i++)
+			fputc(spec->fill, fp);
+	}
+	fprintf(fp, "%s", str);
+	if (spec->align == '<' && spec->width > (unsigned int)len) {
+		for (unsigned i = 0; i < spec->width - len; i++)
+			fputc(spec->fill, fp);
+	}
+	return 0;
+}
+
+int
+xbps_fmt_number(const struct xbps_fmt_spec *spec, int64_t d, FILE *fp)
```

Comment:
would octal also make sense here? might be nice if you want to print numerical file mode

Reply:
maybe there's also some kind of special case that could be added later for doing strmode() on the value? (thinking about how i could use this for `query --long -f`)

Reply:
Switched the format to `{var!humanize[ ][.][width][minscale[maxscale]][i][:spec]}`, this leaves space for `{var!strmode[:spec]}` and allows to configure the conversions independent from the later string padding using the normal spec suffix for variables.
The default for just `!humanize` matches #445. 
```
$ xbps-query -F '{installed_size!humanize:8}\n' -m | head
  79 MiB
 398 KiB
 898 KiB
 1.3 MiB
 0.0 KiB
 1.8 MiB
 6.0 MiB
  72 KiB
  83 KiB
 127 KiB
```

Reply:
nice! maybe short versions of the conversions could also be done? like `!h` for `!humanize` and `!m` for `!strmode`